lib/spack/spack/util/environment.py

```diff
@@ -1075,33 +1075,42 @@ def environment_after_sourcing_files(
             only when the previous command succeeds (default: ``&&``)
     """
     # Set the shell executable that will be used to source files
-    shell_cmd = kwargs.get("shell", "/bin/bash")
-    shell_options = kwargs.get("shell_options", "-c")
-    source_command = kwargs.get("source_command", "source")
-    suppress_output = kwargs.get("suppress_output", "&> /dev/null")
+    if sys.platform == "win32":
+        shell_cmd = kwargs.get("shell", "cmd.exe")
+        shell_options = kwargs.get("shell_options", "/C")
+        suppress_output = kwargs.get("suppress_output", "")
+        source_command = kwargs.get("source_command", "")
+    else:
+        shell_cmd = kwargs.get("shell", "/bin/bash")
+        shell_options = kwargs.get("shell_options", "-c")
+        suppress_output = kwargs.get("suppress_output", "&> /dev/null")
+        source_command = kwargs.get("source_command", "source")
     concatenate_on_success = kwargs.get("concatenate_on_success", "&&")
 
-    shell = Executable(" ".join([shell_cmd, shell_options]))
+    shell = Executable(shell_cmd)
 
     def _source_single_file(file_and_args, environment):
-        source_file = [source_command]
-        source_file.extend(x for x in file_and_args)
-        source_file = " ".join(source_file)
+        source_file_arguments = [shell_options]
+
+        in_bash_args = [source_command]
```

Comment:
`source` needs to be the first thing in the args, if it is used, but this extends that arg list later on with `source_file_arguments.append(" ".join(in_bash_args))`. For the time being, I think it would be safer to have a high-level Windows-vs.-elsewhere branch, and we can consolidate in a follow-up PR.